### PR TITLE
[Deprecated] Training & evaluation of GCN classifier 🏷️ 

### LINKED ISSUE #147: [BUG]: Unable to train the model successfully
### Describe the bug

Training doesn't seem to proceed well - accuracy saturates after a few epochs. Accuracy & confusion matrix do not match. 

Training log [2023-08-22_17-34-27]:

Screenshot 2023-08-22 at 17 48 38

Similarly, the confusion matrices always look like below:

Screenshot 2023-08-23 at 11 38 36

### LINKED ISSUE #186: Pre-process predictions so that only `.long()` dtype vectors are passed into metrics
Metrics are not updated correctly as there's a bug in the metrics predictions vector pre-processing.

### LINKED ISSUE #187: Include a `self.predict()` method into the GCN classifier
Run `predict` for inference, setting dropout to 0.0.

### LINKED ISSUE #188: Run model inference on random graph + real annotated data
Write a notebook, to start with.

### LINKED ISSUE #190: [BUG] Fix how the subgraph dataset is created at inference time
Current code throws errors if `mode == "whole"` is selected because `sub_graph` is not defined (and shouldn't be):

https://github.com/alan-turing-institute/grace/blob/e634d3621747d975f974aade96141bb9174d5594/grace/models/datasets.py#L81-L92

### LINKED ISSUE #193: Rename `test` to `valid` function in `train.py`

### LINKED ISSUE #194: Measure area under [receiver-operator, precision-recall] curve metrics
Perform more elaborate GCN evaluation at inference time.

### LINKED ISSUE #198: Avoid the matrix shape crash in tests - dynamically change embedding num_dims
Should there be some fix to avoid the crash we are getting in the tests?  I would expect the size of the embeddings might change with the input datasets or as we improve the training, so the code should be able to dynamically change the number of dimensions of the embedding without crashing (read it from the config file).

_Originally posted by @crangelsmith in https://github.com/alan-turing-institute/grace/issues/183#issuecomment-1705034215_

### LINKED ISSUE #202: Shuffle the dataset before train-to-valid split
To avoid training / validating on data with unwanted correlations / relationships, _e.g._ on one single line.

### LINKED ISSUE #208: Allow empty `hidden_channels` list config hyperparam to create model with no GC layers
Currently, tests don't pass when GC layers are commented out / not supplied.

### PR DESCRIPTION
**Pull request by KristinaUlicna**
Debugged the node & edge classification with multiple issues. 

When reviewing, try these config hyperparameters on a simple image: 
```
    "patch_size": "(224, 224)",
    "keep_patch_fraction": "1.0",
    "feature_dim": "2048",
    "num_node_classes": "2",
    "num_edge_classes": "2",
    "epochs": "500",
    "hidden_channels": "[]",
    "metrics": "['accuracy', 'confusion_matrix']",
    "dropout": "0.0",
    "batch_size": "512",
    "learning_rate": "0.1",
    "tensorboard_update_frequency": "1"
```

Issues addressed to debug the classifier:
- [ ] Resolves #147 
- [ ] Resolves #190
- [ ] Resolves #193 
- [ ] Resolves #202 
- [ ] Resolves #186 
- [ ] Resolves #208
- [ ] Resolves #198 

**Split the visualisations & enhancements:**
- [ ] Resolves #187
- [ ] Resolves #188
- [ ] Resolves #194

### CODE REVIEW

**Comment by KristinaUlicna** [comment]
Tests for model build fail as I've removed the GCN layers while debugging. WIP 🚧 👷‍♀️

**Comment by crangelsmith** [comment]
Thanks @KristinaUlicna, I'm starting the review now. Just to understand, is this PR also addressing #147, or are you still not satisfied with the solution? I don't see the issue referenced above.

**Comment by KristinaUlicna** [comment]
@crangelsmith Well spotted, I'm adding the #147 in now. The model should train (considerably) well now 🚀

File: grace/models/classifier.py [hunk 1]
```diff
@@ -44,6 +44,8 @@ def __init__(
         # Hidden channels start from input features:
         hidden_channels_list = [input_channels] + hidden_channels
 
+        # Define the list of graph convolutional layers:
+        self.conv_layer_list = []
```
Comment (crangelsmith) [inline grace/models/classifier.py:48]: is there a reason to have this first declaration before line 48?

File: grace/models/classifier.py [hunk 2]
```diff
@@ -60,24 +62,90 @@ def forward(
...
+            Logit predictions of each node class.
+        edge_x : torch.Tensor
+            Logit predictions of each edge class.
+        """
+        # Ensure the model is in training mode:
+        self.train()
+
+        # Run through a series of graph convolutional layers:
+        # for layer in range(len(self.conv_layer_list)):
+        #     x = self.conv_layer_list[layer](x, edge_index)
```
Comment (crangelsmith) [inline grace/models/classifier.py:91]: do we think we will revert back to this commented code? if not why keeping it?

File: grace/models/feature_extractor.py [hunk 3]
```diff
@@ -173,14 +173,27 @@ def forward(
...
 
-            bbox_image = self.normalize_func(bbox_image)
+            # bbox_image = self.normalize_func(bbox_image)
+
+            # Extract the features from ResNet-152 model:
             features = self.model(bbox_image)
             node_attrs[GraphAttrs.NODE_FEATURES] = features.squeeze()
 
+            # HACK:
+            # Average the 4x4 central pixel values of each node patch:
```
Comment (crangelsmith) [inline grace/models/feature_extractor.py:188]: same question as above, why keeping these comments?

File: grace/models/classifier.py [hunk 4]
```diff
@@ -60,24 +62,90 @@ def forward(
...
+        # Ensure the model is in evaluation mode:
+        self.eval()
+
+        # Forward pass through the model
+        with torch.no_grad():
+            node_emb, edge_emb, node_x, edge_x = self(x, edge_index)
+            # predicted_classes = torch.argmax(node_x, dim=-1)
 
-        return node_x, edge_x
+        return node_emb, edge_emb, node_x, edge_x
```
Comment (crangelsmith) [inline grace/models/classifier.py:151]: Just to understand, why are we returning now node_emb, edge_emb? Looks like is not used anywhere and is part of the input right?

**Comment by crangelsmith** [comment]
Should there be some fix to avoid the crash we are getting in the tests?  I would expect the size of the embeddings might change with the input datasets or as we improve the training, so the code should be able to dynamically change the number of dimensions of the embedding without crashing (read it from the config file).

**Comment by KristinaUlicna** [comment]
Thanks @crangelsmith for a thorough feedback & catch up earlier today 🌟  All your suggestions are implemented now (outdating the comments), tests are passing as I've enabled the classifier to accept no hidden layers. I've split the PR into multiple smaller contributions, will continue to separate the training from the evaluation parts 🚀

File: grace/models/feature_extractor.py (hunk 3 above)
Comment (KristinaUlicna) [reply grace/models/feature_extractor.py:188]: I'm not sure how to approach this - when testing different configs & hyperparameters, it may be useful to keep an option to just use the central patch average intensity, which is informative for the synthetic dataset. It's much faster compared to running each patch through an entire Resnet-152, which takes about 30-40s per single image. What do you reckon is the best to do here?

**Comment by KristinaUlicna** [comment]
Deprecated as I have split the contribution into smaller PRs in the meantime. All code is now effectively in `main` 🚀
